controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,410 @@
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	if err := r.reconcileFloatingIPNetwork(scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+		if !claim.ObjectMeta.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			ipAddress := &ipamv1.IPAddress{}
+			if err := r.Client.Get(ctx, client.ObjectKey{Name: claim.Name, Namespace: claim.Namespace}, ipAddress); err == nil {
+				// IPAddress already exists, another reconciler is working on it
+				log.Info("IPAddress already exists, another reconciler is working on it")
+				continue
+			}
+
+			ip, err := r.getIP(scope, pool)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress = &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+					Prefix:  32,
+				},
+			}
+
+			// If the ReclaimPolicy is Delete, we add the finalizer to the IPAddress if they are not pre-allocated
+			// this is to ensure that the IPAddress is deleted from openstack when the claim is deleted
+			if pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete && !contains(pool.Spec.PreAllocatedFloatingIPs, ip) {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
+			if err = r.Client.Status().Update(ctx, &claim); err != nil {
+				log.Error(err, "Failed to update IPAddressClaim status", "claim", claim.Name, "ip", ip)
+				return ctrl.Result{}, err
+			}
+			scope.Logger().Info("Claimed IP", "ip", ip)
+		}
+	}
+	if err = r.setIPStatuses(ctx, pool); err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
This is redundant: any change which could have resulted in a change to IP statuses will result in another reconcile, in which case the first thing we will do is call setIPStatuses(). We don't need to do this twice in the same reconcile.